substrate/utils/frame/benchmarking-cli/src/overhead/command.rs

```diff
@@ -0,0 +1,773 @@
+		>,
+		chain_spec: Option<Box<dyn ChainSpec>>,
+	) -> Result<()>
+	where
+		Block: BlockT<Extrinsic = OpaqueExtrinsic>,
+		ExtraHF: HostFunctions,
+	{
+		if let Err((error_kind, msg)) = self.check_args(&chain_spec) {
+			let mut cmd = OverheadCmd::command();
+			cmd.error(error_kind, msg).exit();
+		};
+
+		let (state_handler, para_id) =
+			self.state_handler_from_cli::<(ParachainHostFunctions, ExtraHF)>(chain_spec)?;
+
+		let executor = WasmExecutor::<(ParachainHostFunctions, ExtraHF)>::builder()
+			.with_allow_missing_host_functions(true)
+			.build();
+
+		let metadata =
+			fetch_latest_metadata_from_code_blob(&executor, state_handler.get_code_bytes()?)?;
+
+		// At this point we know what kind of chain we are dealing with.
+		let chain_type = identify_chain(&metadata, para_id);
+
+		// If we are dealing  with a parachain, make sure that the para id in genesis will
+		// match what we expect.
+		let genesis_patcher = match chain_type {
+			Parachain(para_id) =>
+				Some(Box::new(move |value| patch_genesis(value, Some(para_id))) as Box<_>),
+			_ => None,
+		};
+
+		let client = self.build_client_components::<Block, (ParachainHostFunctions, ExtraHF)>(
+			state_handler.build_storage::<(ParachainHostFunctions, ExtraHF)>(genesis_patcher)?,
+			executor,
+			&chain_type,
+		)?;
+
+		let inherent_data = create_inherent_data(&client, &chain_type);
+
+		let (ext_builder, runtime_name) = {
+			let genesis = client.usage_info().chain.best_hash;
+			let version = client.runtime_api().version(genesis).unwrap();
+			let runtime_name = version.spec_name;
+			let runtime_version = RuntimeVersion {
+				spec_version: version.spec_version,
+				transaction_version: version.transaction_version,
+			};
+
+			(ext_builder_provider(metadata, genesis, runtime_version), runtime_name)
+		};
+
+		self.run(
+			runtime_name.to_string(),
+			client,
+			inherent_data,
+			Default::default(),
+			&*ext_builder,
+			chain_type.requires_proof_recording(),
+		)
+	}
+
+	/// Run the benchmark overhead command.
+	pub fn run_with_extrinsic_builder<Block, ExtraHF>(
+		&self,
+		ext_builder_provider: Box<
+			dyn FnOnce(Metadata, Block::Hash, RuntimeVersion) -> Box<dyn ExtrinsicBuilder>,
+		>,
+	) -> Result<()>
+	where
+		Block: BlockT<Extrinsic = OpaqueExtrinsic>,
+		ExtraHF: HostFunctions,
+	{
+		self.run_with_extrinsic_builder_and_spec::<Block, ExtraHF>(ext_builder_provider, None)
+	}
+
+	fn build_client_components<Block, HF>(
+		&self,
+		genesis_storage: Storage,
+		executor: WasmExecutor<HF>,
+		chain_type: &ChainType,
+	) -> Result<Arc<OverheadClient<Block, HF>>>
+	where
+		Block: BlockT,
+		HF: HostFunctions,
+	{
+		let extensions = ExecutionExtensions::new(None, Arc::new(executor.clone()));
+
+		let base_path = match &self.shared_params.base_path {
+			None => BasePath::new_temp_dir()?,
+			Some(path) => BasePath::from(path.clone()),
+		};
+
+		let database_source = self.database_config(
+			&base_path.path().to_path_buf(),
```

Comment:
So we can already run it on a state snapshot to get real PoV size? Nice.

Reply:
I think it should work, but did not test that tbh. I was pretty focused on making the base case work first. We should follow up on it.